include/material/mohr_coulomb.tcc

```diff
@@ -0,0 +1,571 @@
+  const double G = shear_modulus_;
+  const double a1 = bulk_modulus_ + (4.0 / 3.0) * G;
+  const double a2 = bulk_modulus_ - (2.0 / 3.0) * G;
+  // compute elastic stiffness matrix
+  // clang-format off
+  de_(0,0)=a1;    de_(0,1)=a2;    de_(0,2)=a2;    de_(0,3)=0;    de_(0,4)=0;    de_(0,5)=0;
+  de_(1,0)=a2;    de_(1,1)=a1;    de_(1,2)=a2;    de_(1,3)=0;    de_(1,4)=0;    de_(1,5)=0;
+  de_(2,0)=a2;    de_(2,1)=a2;    de_(2,2)=a1;    de_(2,3)=0;    de_(2,4)=0;    de_(2,5)=0;
+  de_(3,0)= 0;    de_(3,1)= 0;    de_(3,2)= 0;    de_(3,3)=G;    de_(3,4)=0;    de_(3,5)=0;
+  de_(4,0)= 0;    de_(4,1)= 0;    de_(4,2)= 0;    de_(4,3)=0;    de_(4,4)=G;    de_(4,5)=0;
+  de_(5,0)= 0;    de_(5,1)= 0;    de_(5,2)= 0;    de_(5,3)=0;    de_(5,4)=0;    de_(5,5)=G;
+  // clang-format on
+
+  return true;
+}
+
+//! Compute stress invariants
+template <unsigned Tdim>
+bool mpm::MohrCoulomb<Tdim>::compute_stress_invariants(
+    const Vector6d& stress, mpm::dense_map* state_vars) {
+  // Compute the mean pressure
+  const double mean_p = (stress(0) + stress(1) + stress(2)) / 3.;
+  // Compute the deviatoric stress
+  Vector6d dev_stress = stress;
+  for (unsigned i = 0; i < 3; ++i) dev_stress(i) -= mean_p;
+  // Compute J2
+  double j2 = (std::pow((stress(0) - stress(1)), 2) +
+               std::pow((stress(1) - stress(2)), 2) +
+               std::pow((stress(0) - stress(2)), 2)) /
+                  6.0 +
+              std::pow(stress(3), 2) + std::pow(stress(4), 2) +
+              std::pow(stress(5), 2);
+  // Compute J3
+  (*state_vars).at("j3") = dev_stress(0) * dev_stress(1) * dev_stress(2) -
+                           dev_stress(2) * std::pow(dev_stress(3), 2) +
+                           2 * dev_stress(3) * dev_stress(4) * dev_stress(5) -
+                           dev_stress(0) * std::pow(dev_stress(4), 2) -
+                           dev_stress(1) * std::pow(dev_stress(5), 2);
+  // Compute theta value (Lode angle)
+  double theta_val = 0.;
+  if (fabs(j2) > 0.0)
+    theta_val =
+        (1.5 * std::sqrt(3.)) * ((*state_vars).at("j3") / std::pow(j2, 1.5));
+  // Check theta value
+  if (theta_val > 1.0) theta_val = 1.0;
+  if (theta_val < -1.0) theta_val = -1.0;
+  // Compute theta
+  (*state_vars)["theta"] = (1. / 3.) * acos(theta_val);
+  // Check theta
+  if ((*state_vars).at("theta") > M_PI / 3.) (*state_vars)["theta"] = M_PI / 3.;
+  if ((*state_vars).at("theta") < 0.0) (*state_vars)["theta"] = 0.;
+  // Compute rho
+  (*state_vars)["rho"] = std::sqrt(2. * (j2));
+  // Compute epsilon
+  (*state_vars)["epsilon"] =
+      (1. / std::sqrt(3.)) * (stress(0) + stress(1) + stress(2));
+
+  return true;
+}
+
+//! Compute yield function and yield state
+template <unsigned Tdim>
+typename mpm::mohrcoulomb::FailureState
+    mpm::MohrCoulomb<Tdim>::compute_yield_state(
+        Eigen::Matrix<double, 2, 1>* yield_function,
+        const mpm::dense_map& state_vars) {
+  // Tolerance for yield function
+  const double Tolerance = -1E-1;
```

Comment:
Following @ezrayst comment, this one too. Isn't it to large?

Reply:
@tianchiTJ this part is not used for iteration, so we can make it more accurate.